ukbb_qc/release/prepare_vcf_data_release.py

```diff
@@ -601,121 +650,222 @@ def main(args):
-                    ht._map_partitions(
-                        lambda p: hl.array([part_min_and_max(p)])
-                    ).collect()
-                )
+            if args.test:
+                mt = mt._filter_partitions(range(2))
+            mt = hl.filter_intervals(mt, [hl.parse_locus_interval(contig)])
+            mt = mt.annotate_rows(**ht[mt.row_key])
+
+            logger.info("Adjusting partitions...")
+            mt = mt.naive_coalesce(n_partitions)
+            logger.info("%s has %i partitions", contig, mt.n_partitions())
+
+            logger.info("Densifying...")
+            mt = hl.experimental.densify(mt)
+            # Drop END and het non ref to avoid exporting
+            mt = mt.drop("het_non_ref")
+
+            logger.info("Removing low QUAL variants and * alleles...")
+            info_ht = hl.read_table(info_ht_path(data_source, freeze))
+            mt = mt.filter_rows(
+                (~info_ht[mt.row_key].AS_lowqual)
+                & ((hl.len(mt.alleles) > 1) & (mt.alleles[1] != "*"))
+            )
+
+            logger.info("Adjusting partitions again post-densify...")
+            mt = mt.naive_coalesce(n_partitions)
+            logger.info("%s has %i partitions post-densify", contig, mt.n_partitions())
+            ht = mt.rows()
+            # Unkey HT to avoid this error with map_partitions:
+            # ValueError: Table._map_partitions must preserve key fields
+            ht = ht.key_by()
+
+            logger.info("Adjusting sex ploidy...")
+            mt = adjust_sex_ploidy(mt, mt.sex_karyotype, male_str="XY", female_str="XX")
+            mt = mt.select_cols()
+
+            logger.info("Exporting VCF...")
+            hl.export_vcf(
+                mt,
+                release_vcf_path(*tranche_data, contig=contig),
+                metadata=header_dict,
+                append_to_header=append_to_vcf_header_path(*tranche_data),
+                parallel="header_per_shard",
+                # Removed tabix here because the repackaging code produces an index
+            )
+
+            logger.info("Getting start and stops per shard...")
+
+            def part_min_and_max(part):
+                keys = part.map(lambda x: x.select("locus", "alleles"))
+                return hl.struct(start=keys[0], end=keys[-1])
+
+            start_stop_list = ht._map_partitions(
+                lambda p: hl.array([part_min_and_max(p)])
+            ).collect()
+            print(start_stop_list)
+            with hl.hadoop_open(
+                f"gs://broad-ukbb/broad.freeze_7/release/vcf_positions_tsvs/{contig}_start_end_pos.tsv",
+                "w",
+            ) as o:
+                o.write("vcf_shard_name\tstart_pos\tend_pos\n")
+                for count, struct in enumerate(start_stop_list):
+                    # Format shard name to have leading zeros
+                    # e.g., 0 should be part-000000.bgz
+                    shard_name = f"part-{count:05d}.bgz"
```

Comment:
Cool -- I've never had to do this before in Hail, so I learned this reading your code!